### ISSUE TITLE
Blurry SVG ImageStatic after 7.3.0

### ISSUE THREAD

**Issue by MichalK6677** [openlayers/openlayers#14788]
LayerImage with ImageStatic with SVG image is blurry after update from 7.2.2 to 7.3.0 and later.
It seems to be related with this: #14444 

**See the difference:**
### **7.2.2**
https://codesandbox.io/s/openlayers-7-2-2-svg-issue-hrwp1z?file=/index.js
![ol-7 2 2](https://github.com/openlayers/openlayers/assets/3100059/dbd9f6d9-3284-4eee-a91d-5d568b21f4c9)

### **7.3.0**
https://codesandbox.io/s/openlayers-7-3-0-svg-issue-2xrwrf?file=/index.js
![ol-7 3 0](https://github.com/openlayers/openlayers/assets/3100059/aa075aaf-404c-4f0d-b04b-b293a5c9ef16)

You can try to uncoment section below, it fixes the issue:
![image](https://github.com/openlayers/openlayers/assets/3100059/aa054fd8-0ce7-4af8-b5c4-6d837847b1d0)

How can I achieve behaviour of 7.2.2 version without overring OL functions?

**Comment by ahocevar** [comment]
This behaviour was indeed introduced with #14444. I agree it is not ideal for your use case. And I don't have a good solution how the library could serve your use case well with an `ImageStatic` layer after #14444. Maybe others in @openlayers/core have ideas?

Have you considered creating a barebone layer with a custom renderer like shown in https://openlayers.org/en/latest/examples/svg-layer.html? Since you have render customizations with `prerender` and `postrender` anyway, I think that solution would be more appropriate.

**Comment by mike-000** [comment]
Some improvement in the current code is possible by upscaling the image when it is drawn to canvas, for example

```
var upScale = 16;
var overlay = {
  LongLat1: [18.73514003, 49.23716287],
  Asset: {
    Url: "/floor1.svg",
    Scale: 0.9056349551449104 / upScale,
    Resolution: 0.03732276771737122,
    Rotation: -0.9501912598404259,
    Width: 640 * upScale,
    Height: 494 * upScale,
    Opacity: 1
  }
};
```


For a more general solution a `useImage` option in the ImageStatic constructor could be used to prevent a context being passed to ImageWrapper https://codesandbox.io/s/openlayers-7-3-0-svg-issue-forked-fekw46?file=/index.js

**Comment by ahocevar** [comment]
I'd like to avoid adding options to the API that are implementation or renderer specific.

**Comment by MichalK6677** [comment]
I can play with the layer with a custom renderer but I see in the example that the SVG is created as a separate HTML node, it's not a part of canvas. Is it possible to apply zIndexes on this SVG mixed with layers directly in canvas? 

I mean something like TileLayer (zIndex: 1), custom renderer layer with SVG (zIndex: 2), Vector layer with vector source with points on map (zIndex: 3).

**Comment by ahocevar** [comment]
Yes, layer order (including zIndex) will be respected as usual, when you have a custom renderer in between other layers.

**Comment by MichalK6677** [comment]
I kinda like also solution from @mike-000 with the upscaling. It could solve also performance issues with very complex SVGs as they are rasterized and upscaled.

**Comment by ahocevar** [comment]
Oh yes, performance wise, if you're not short on memory, upscaling the SVG would be the best solution.

**Comment by mike-000** [comment]
Note that your method for setting the size of an SVG without width and height attributes does not work consistently across all browsers (it will not work at all with Firefox in any OpenLayers version), similar to the issues discussed in #14196.  A more reliable approach would be to fetch the svg and set the attributes in a data url

    imageLoadFunction: (image, src) => {
      fetch(src)
        .then((response) => response.text())
        .then((text) => {
          const xml = new DOMParser().parseFromString(text, 'image/svg+xml');
          const svg = xml.getElementsByTagName('svg')[0];
          svg.setAttribute('width', overlay.Asset.Width);
          svg.setAttribute('height', overlay.Asset.Height);
          const src =
            'data:image/svg+xml,' +
            encodeURIComponent(new XMLSerializer().serializeToString(xml));
          image.getImage().src = src;
        });
    },


Before version 7.3.0 I think the `imageSize` option would have had a similar (inconsistent) effect to you method, but it is no longer useful and could be deprecated.

**Comment by MichalK6677** [comment]
I'm playing a bit with SVG layer and zIndex and seems there is something wrong:

https://codesandbox.io/s/openlayers-7-3-0-svg-issue-forked-37e9yq?file=/index.js

Try to zoom in and out and check the red circle. It does not stay on its position. It's moving. When I change zIndex of the svgLayer to e.g. 6 (svgLayer is on top), then position of the red circle is ok.

### svgLayer zIndex = 4
![zIndex4](https://github.com/openlayers/openlayers/assets/3100059/b8c9c529-972e-4ca3-807a-6bdd9885b26e)

### svgLayer zIndex = 6
![zIndex6](https://github.com/openlayers/openlayers/assets/3100059/d84b05d0-c34b-4041-b33e-44d72fce9626)

**Comment by mike-000** [comment]
Adding `svgContainer.style.position = "absolute";` will fix that.

**Comment by MichalK6677** [comment]
Is this the correct way how to take advantage of #14945 ?
https://codesandbox.io/s/openlayers-7-3-0-svg-issue-forked-fpjkdq

Seems it works similar way like 7.2.2 (SVG is supported again) but it still reloads the image with every pan or zoom.

**Comment by mike-000** [comment]
> Is this the correct way how to take advantage of #14945 ?

Yes, it draws the image to the rendering canvas at each map move.  That could be slow and/or cause a noticeable flicker depending on the browser and the complexity of the SVG.  In this case it seems much worse on Chrome than Firefox, although if is not correctly scaled on Firefox unless I insert `preserveAspectRatio="none"` into the SVG header.  The difference is surprising as in other tests I found Firefox to be much slower than Chrome.

**Comment by ahocevar** [comment]
@MichalK6677 Where do you see the image reload with every pan or zoom?

**Comment by mike-000** [comment]
> @MichalK6677 Where do you see the image reload with every pan or zoom?

It does on latest Chrome on Windows 10 for me.

**Comment by MichalK6677** [comment]
I see the image is flickering and I see the reload in DevTools. This can take significant data load in my case when I use more images.
![svgTest](https://github.com/openlayers/openlayers/assets/3100059/80936e67-5e1f-456d-9d0b-2f6652ec6ce1)

**Comment by ahocevar** [comment]
@MichalK6677 This is because something in your setup does not appear to be valid (I assume the SVG you're using). In a correct setup, the image should not wander around. You can set the image extent also on the layer as `extent`, which will stop the image from loading over and over, and also show the extent where the image should always be rendered - which it isn't.

**Comment by MichalK6677** [comment]
@ahocevar Is there any working example for `createLoader from "ol/source/static.js"` with static SVG?

**Comment by mike-000** [comment]
Your code is correct.  See https://deploy-preview-15013--ol-site.netlify.app/en/latest/examples/reprojection-image-svg.html for a working example (except that is reprojected, which does not work in the dev version).  If I use the url from that example in your code it works smoothly, so I think Chrome does not like your SVG - although apart from the known issue with aspect ratio Firefox has no problem with it.

**Comment by ahocevar** [comment]
The codesandbox that @mike-000 linked cannot be compared, because it reprojects the SVG image. For a codesandbox with an unprojected SVG, see https://codesandbox.io/s/reprojection-image-forked-22qhth?file=/main.js.

And strangely, I see the same issue in Chrome - it does indeed get reloaded. I'm going to fix that in a separate pull request.

**Comment by mike-000** [comment]
I think this might be a CodeSandbox problem.  If I upload to CodeSandbox a file which is working correctly when accessed from a remote source there the problem is seen when it is accessed locally in the CodeSandbox.

**Comment by ahocevar** [comment]
No, it is a problem in the library. I'm on it.

**Comment by ahocevar** [comment]
#15039 will fix this.
